### ISSUE TITLE
[BUG] Deleting invoice books back money twice

### ISSUE THREAD

**Issue by CodeNamedRobin** [GEWIS/sudosos-backend#249]
Description: Basically title, but whenever an invoice is deleted on an account, the money is booked back twice. This might only happen when the invoice is a credit invoice, but I'm not too sure. It does create two transfers, both with description "Deletion of Invoice #XXX"

**Comment by tomudding** [comment]
Just to add some more documentation, first _undo_ transfer is created here (introduced in b643a0bab8ead443e81ede965610f484d7a6d46e):
https://github.com/GEWIS/sudosos-backend/blob/024153e3435438cb498122775897c1d325f78eb6/src/service/invoice-service.ts#L347-L360
The other one:
https://github.com/GEWIS/sudosos-backend/blob/024153e3435438cb498122775897c1d325f78eb6/src/service/invoice-service.ts#L268-L273

**Comment by JustSamuel** [comment]
Mhmm, interesting. I’ll try and fix this tomorrow, but for now try and refrain from deleting invoices.Sent from my iPhoneOn 9 Aug 2024, at 22:26, Tom Udding ***@***.***> wrote:﻿
Just to add some more documentation, first undo transfer is created here (introduced in b643a0b):
https://github.com/GEWIS/sudosos-backend/blob/024153e3435438cb498122775897c1d325f78eb6/src/service/invoice-service.ts#L347-L360
The other one:
https://github.com/GEWIS/sudosos-backend/blob/024153e3435438cb498122775897c1d325f78eb6/src/service/invoice-service.ts#L268-L273

**Comment by tomudding** [comment]
> Mhmm, interesting. I’ll try and fix this tomorrow, but for now try and refrain from deleting invoices.

My comment is only to show that two transfers are created for a deletion, not that it is faulty code. The first one is from the void to something/someone, and the second is from someone/something to the void.

**Comment by JustSamuel** [comment]
Invoices always have two parties, the seller and the buyer, so there are always two transfers.Sent from my iPhoneOn 10 Aug 2024, at 15:08, Tom Udding ***@***.***> wrote:﻿

Mhmm, interesting. I’ll try and fix this tomorrow, but for now try and refrain from deleting invoices.

My comment is only to show that two transfers are created for a deletion, not that it is faulty code. The first one is from the void to something/someone, and the second is from someone/something to the void.

**Comment by JustSamuel** [comment]
Still seems to be an issue with credit invoices, will try and make a specific test case for it.

**Comment by JustSamuel** [comment]
Not entirely sure what the main issue is, but a bug has been confirmed here:  https://github.com/GEWIS/sudosos-backend/commit/dc9adf8bf21dbce8392949ebc60588f0a98a27d9

(Although this makes it look like _no_ money would be returned.)

**Comment by JustSamuel** [comment]
Found the issue:

When a invoice is created the invoiced amount is subtracted from the seller's party balance. This is because the seller will now receive this invoice by bank transfer and this money goes "out" of SudoSOS. This makes sense, because creating an invoice is like registering sales in SudoSOS but saying that the payment goes apart form SudoSOS.

When an invoice is deleted the money goes back to the sellers. However in the case of a credit Invoice there are no real "buyers" and the seller account gets
- the transfer returning the money from the invoice as usual
- the seller return invoice

However in the case of a credit invoice only one should be created. So it was indeed correct that SudoSOS accidentally created two transfers. However, in most cases, this should have lead to a net difference of 0 since money would be mistakenly subtracted, but also paid back.

So how the I.V.V account got a net positive deletion is still strange

**Comment by JustSamuel** [comment]
> So how the I.V.V account got a net positive deletion is still strange

I found it!

I.V.V had a 100 in sales. Create a credit invoice of 100 and the balance is now 0.
"Delete" the invoice and the balance stays at 0
Create a new invoice of 100 and I.V.V is now at -100

The only thing that I am uncertain of is that there was a bug where a Invoice was created twice but there was a error during the creation. So the balance was subtracted even further without a prior invoice attached.

The transfers have been removed and the balance is now back at 0
